File: api/slack-bot-processor/src/processors/document.ts
import pdf from 'pdf-parse'
import mammoth from 'mammoth'
import { registerProcessor } from './registry.js'
import type { ProcessorInput, ProcessorOutput } from './types.js'

/**
 * PDF processor - extract text from PDFs
 */
async function processPdf(input: ProcessorInput): Promise<ProcessorOutput> {
  const data = await pdf(input.buffer)

  const preview = data.text.slice(0, 1000)
  const truncated = data.text.length > 1000 ? '...(truncated)' : ''

  return {
    text: `*PDF Extracted: ${input.filename}*\nPages: ${data.numpages}\n\n\`\`\`\n${preview}${truncated}\n\`\`\``,
    replyInThread: true,
  }
}

/**
 * Word document processor - extract text from .docx
 */
async function processDocx(input: ProcessorInput): Promise<ProcessorOutput> {
  const result = await mammoth.extractRawText({ buffer: input.buffer })

  const preview = result.value.slice(0, 1000)
  const truncated = result.value.length > 1000 ? '...(truncated)' : ''

  return {
    text: `*Document Extracted: ${input.filename}*\n\n\`\`\`\n${preview}${truncated}\n\`\`\``,
    replyInThread: true,
  }
}

// Register PDF processor
registerProcessor({
  extensions: ['pdf'],
  mimetypes: ['application/pdf'],
  process: processPdf,
  description: 'Extract text from PDF files',
})

// Register Word document processor
registerProcessor({
  extensions: ['docx'],
  mimetypes: [
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  ],
  process: processDocx,
  description: 'Extract text from Word documents',
})

File: bot/slack-processor-python-node/src/processors/document.py
"""Document processors - extract text from PDF and DOCX."""

from io import BytesIO

from pypdf import PdfReader
from docx import Document

from .registry import register_processor
from .types import ProcessorInput, ProcessorOutput


@register_processor(
    extensions=["pdf"],
    mimetypes=["application/pdf"],
    description="Extract text from PDF files",
)
def process_pdf(input: ProcessorInput) -> ProcessorOutput:
    """Extract text from PDF."""
    reader = PdfReader(BytesIO(input.content))

    text_parts = []
    for page in reader.pages:
        text_parts.append(page.extract_text() or "")

    full_text = "\n".join(text_parts)
    preview = full_text[:1000]
    truncated = "...(truncated)" if len(full_text) > 1000 else ""

    return ProcessorOutput(
        text=f"*PDF Extracted: {input.filename}*\nPages: {len(reader.pages)}\n\n```\n{preview}{truncated}\n```",
        reply_in_thread=True,
    )


@register_processor(
    extensions=["docx"],
    mimetypes=["application/vnd.openxmlformats-officedocument.wordprocessingml.document"],
    description="Extract text from Word documents",
)
def process_docx(input: ProcessorInput) -> ProcessorOutput:
    """Extract text from DOCX."""
    doc = Document(BytesIO(input.content))

    text_parts = [para.text for para in doc.paragraphs]
    full_text = "\n".join(text_parts)

    preview = full_text[:1000]
    truncated = "...(truncated)" if len(full_text) > 1000 else ""

    return ProcessorOutput(
        text=f"*Document Extracted: {input.filename}*\n\n```\n{preview}{truncated}\n```",
        reply_in_thread=True,
    )
